### PR TITLE
Add battery service

### PR DESCRIPTION
**Pull request by Markushaldorsen**

### CODE REVIEW

File: adapter/service/battery/service.go [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+
+	s.reportingCache.Reported(EvtLevelReport, "", level)
+
+	return true, nil
+}
+
+// SendBatteryAlarmReport sends a battery alarm report. Returns true if a report was sent.
+// Depending on a caching and reporting configuration the service might decide to skip a report.
+// To make sure report is being sent regardless of circumstances set the force argument to true.
+func (s *service) SendBatteryAlarmReport(force bool) (bool, error) {
```
Comment (rstraczynski) [inline adapter/service/battery/service.go:137]: I wonder if providing a report should be a responsibility of a controller. What are the scenarios for an alarm? Is the list finite or they may occur some custom scenarios (so custom alarms would be needed). If there's a fixed number of alarms, maybe the service should be responsible for observing a controller and send an alarm if needed? Let me know what you think.

**Review by rstraczynski — CHANGES_REQUESTED** [review]

File: adapter/service/battery/service.go [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+	level, err := s.reporter.BatteryLevelReport()
+	if err != nil {
+		return false, fmt.Errorf("failed to get battery level: %w", err)
+	}
+
+	if !force && !s.reportingCache.ReportRequired(s.reportingStrategy, EvtLevelReport, "", level) {
+		return false, nil
+	}
+
+	message := fimpgo.NewIntMessage(
```
[the remark this answers is not in the document]
Comment (boguslaw-wojcik) [reply adapter/service/battery/service.go:134]: Yes.

File: adapter/service/battery/service.go (hunk 1 above)
Comment (boguslaw-wojcik) [reply adapter/service/battery/service.go:156]: I think we can leave it as it is, however, we should also add some consts. Person implementing should not guess activ/deactiv and low_battery string.

Comment (Markushaldorsen) [reply adapter/service/battery/service.go:163]: Added constants for activ/deactiv/low_battery.

File: adapter/service/battery/routing_test.go [hunk 3]
```diff
@@ -0,0 +1,385 @@
...
+	}
+
+	fullValue2Struct := battery.FullReport{
+		Level:  80,
+		Health: 50,
+		State:  "charging",
+		Temp:   30.5,
+	}
+
+	alarmValue1StrMap := map[string]string{
```
Comment (rstraczynski) [inline adapter/service/battery/routing_test.go:227]: No need to define those maps here - you can call a method on a structs below ;)

File: adapter/service/battery/routing_test.go [hunk 4]
```diff
@@ -0,0 +1,385 @@
...
+		"event":  battery.AlarmLowBatteryEvent,
+		"status": battery.AlarmStatusActivate,
+	}
+
+	alarmValue2StrMap := map[string]string{
+		"event":  battery.AlarmLowBatteryEvent,
+		"status": battery.AlarmStatusDeactivate,
+	}
+
+	alarmValue1Struct := battery.AlarmReport{
```
Comment (rstraczynski) [inline adapter/service/battery/routing_test.go:237]: `alarm1` is enough I think.

File: adapter/service/battery/routing_test.go [hunk 5]
```diff
@@ -0,0 +1,385 @@
...
+		Event:  battery.AlarmLowBatteryEvent,
+		Status: battery.AlarmStatusActivate,
+	}
+
+	alarmValue2Struct := battery.AlarmReport{
```
Comment (rstraczynski) [inline adapter/service/battery/routing_test.go:242]: same as above

File: adapter/service/battery/service.go [hunk 6]
```diff
@@ -0,0 +1,321 @@
...
+	if err != nil {
+		return false, fmt.Errorf("failed to send battery full report: %w", err)
+	}
+
+	s.reportingCache.Reported(EvtBatteryReport, "", full)
+
+	return true, nil
+}
+
+func (bar *AlarmReport) StructToStrMap() (strMap map[string]string, err error) {
```
Comment (rstraczynski) [inline adapter/service/battery/service.go:309]: Please move it above, right below the struct it belongs.

File: adapter/service/battery/service.go [hunk 7]
```diff
@@ -0,0 +1,321 @@
...
+func (s *service) SendBatterySensorReport(force bool) (bool, error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	sensor, unit, err := s.reporter.BatterySensorReport()
+	if err != nil {
+		return false, fmt.Errorf("failed to get battery sensor report: %w", err)
+	}
+
+	props := fimpgo.Props{
```
Comment (rstraczynski) [inline adapter/service/battery/service.go:247]: `props` can be defined below checking if the message should be sent.

**Review by rstraczynski — CHANGES_REQUESTED** [review]

File: adapter/service/battery/service.go (hunk 6 above)
Comment (rstraczynski) [inline adapter/service/battery/service.go:309]: `ToStrMap` will be enough.

File: adapter/service/battery/service.go [hunk 8]
```diff
@@ -0,0 +1,321 @@
...
+		return false, fmt.Errorf("failed to send battery full report: %w", err)
+	}
+
+	s.reportingCache.Reported(EvtBatteryReport, "", full)
+
+	return true, nil
+}
+
+func (bar *AlarmReport) StructToStrMap() (strMap map[string]string, err error) {
+	marshal, err := json.Marshal(bar)
```
Comment (rstraczynski) [inline adapter/service/battery/service.go:310]: I'd return a map directly instead of marshalling/unmarshalling as those are pretty expensive operations.

**Review by rstraczynski — CHANGES_REQUESTED** [review]

File: adapter/service/battery/routing_test.go [hunk 9]
```diff
@@ -0,0 +1,385 @@
...
+	"github.com/futurehomeno/cliffhanger/adapter"
+	"github.com/futurehomeno/cliffhanger/adapter/service/battery"
+	"github.com/futurehomeno/cliffhanger/router"
+	"github.com/futurehomeno/cliffhanger/task"
+	mockedbattery "github.com/futurehomeno/cliffhanger/test/mocks/adapter/service/battery"
+	"github.com/futurehomeno/cliffhanger/test/suite"
+)
+
+func TestRouteBattery(t *testing.T) { // nolint:paralleltest
+	fullValue := battery.FullReport{
```
[the remark this answers is not in the document]
Comment (Markushaldorsen) [reply adapter/service/battery/routing_test.go]: Fixed

File: adapter/service/battery/routing_test.go [hunk 10]
```diff
@@ -0,0 +1,385 @@
...
+				},
+			},
+		},
+	}
+
+	s.Run(t)
+}
+
+func TestTaskBattery(t *testing.T) { // nolint:paralleltest
+	fullValue1Struct := battery.FullReport{
```
[the remark this answers is not in the document]
Comment (Markushaldorsen) [reply adapter/service/battery/routing_test.go]: Fixed

File: adapter/service/battery/routing_test.go [hunk 11]
```diff
@@ -0,0 +1,385 @@
...
+		Level:  90,
+		Health: 60,
+		State:  "charging",
+		Temp:   40.5,
+	}
+
+	fullValue2Struct := battery.FullReport{
```
[the remark this answers is not in the document]
Comment (Markushaldorsen) [reply adapter/service/battery/routing_test.go]: Fixed

File: adapter/service/battery/routing_test.go (hunk 3 above)
Comment (Markushaldorsen) [reply adapter/service/battery/routing_test.go]: Fixed

File: adapter/service/battery/routing_test.go (hunk 4 above)
Comment (Markushaldorsen) [reply adapter/service/battery/routing_test.go]: Fixed

File: adapter/service/battery/service.go (hunk 6 above)
Comment (Markushaldorsen) [reply adapter/service/battery/service.go]: Done

Comment (Markushaldorsen) [reply adapter/service/battery/service.go]: Fixed

**Review by rstraczynski — APPROVED** [review]

**Pull request merged**